zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -222,22 +226,16 @@ protected ZkClient(IZkConnection zkConnection, int connectionTimeout, long opera
     _asyncCallRetryThread.start();
     LOG.debug("ZkClient created with _uid {}, _asyncCallRetryThread id {}", _uid, _asyncCallRetryThread.getId());
 
-    connect(connectionTimeout, this);
-
-    // initiate monitor
-    try {
-      if (monitorKey != null && !monitorKey.isEmpty() && monitorType != null && !monitorType
-          .isEmpty()) {
-        _monitor =
-            new ZkClientMonitor(monitorType, monitorKey, monitorInstanceName, monitorRootPathOnly,
-                _eventThread);
-        _monitor.register();
-      } else {
-        LOG.info("ZkClient monitor key or type is not provided. Skip monitoring.");
-      }
-    } catch (JMException e) {
-      LOG.error("Error in creating ZkClientMonitor", e);
+    if (monitorKey != null && !monitorKey.isEmpty() && monitorType != null && !monitorType
+        .isEmpty()) {
+      _monitor =
+          new ZkClientMonitor(monitorType, monitorKey, monitorInstanceName, monitorRootPathOnly,
+              _eventThread);
```

Comment:
Let's give it a more complete change.
1. The constructor will not require _eventThread anymore, it will be null anyway.
2. setAndInitZkEventThreadMonitor shall create the event thread monitor and call updateAttributesInfo() to update the attribute. Re-register the whole monitor is wrong, you will end up either having 2 mbeans or the 2nd register will fail.

Reply:
Synced with Kai offline, the 2nd point is fine since the register will be postpone to the connect. Given this, I would prefer making the registration logic closer to the monitor construct logic.

So shall we pass the ZkClientMonitor into connect() to explicitly update it in the call. Then register the monitor after the connect(), but still in the constructor.

pseudo code:

_monitor = new ZkClientMonitor();
// Comment that we need to update the _monitor in the event lock to avoid missing any records about event thread.
connect(connectTimeout, this, _monitor);
_monitor.register()

Reply:
This does not seem to work. 

First  it is not necessary to pass _monitor into `connect(connectTimeout, this, _monitor);` as _monitor is instance variable and connect is method. Connect can see it anyway.

Second, the gist to avoid all these racing issue is that we must make sure _monitor is fully constructed and also register-ed before ZkClient either get callback from ZK.

Reply:
1. This is to make the code cleaner. It is a good practice for module the code that the private method does not refer to the class private fields directly.

2. The thread monitor is the only one impacted, so if you do the set there, there should be no problem. Mix everything together is chaotic. Ideally, I would suggest splitting the ZK client monitor and ZK thread monitor.

Reply:
Maybe I mis-understand what do you mean. Can you draw the code a little bit more specific? 

from the above pseduo code

```
_monitor = new ZkClientMonitor();
// Comment that we need to update the _monitor in the event lock to avoid missing any records about event thread.
connect(connectTimeout, this, _monitor);
_monitor.register()
```

construction of ZkMonitor and ZkThreadMonitor and their registration must be done before the first doAsyncSync() is sent out, or you have race condition of missing some read/write.

In the above pseudo code, `_monitor.register()` is too late, the first sycn() is already out and complete earlier than this register().  This violate the above assertion. 

I may not fully understand your idea. Let us be more specific. As long as it works, I can change it to that way.

Reply:
The registration does not need to happen prior. Even not registered, the metric can record data.

Reply:
I test it. It seems not true. We need to register the bean in connect() under the lock. Otherwise, depends on timing, you may lose one read.